src/topology/algebra/nonarchimedean/bases.lean

```diff
@@ -0,0 +1,330 @@
+of neighborhoods of zero. -/
+def topology : topological_space A :=
+hB.to_ring_filter_basis.to_add_group_filter_basis.topology
+
+lemma has_basis_nhds_zero : has_basis (@nhds A hB.topology 0) (λ _, true) (λ i, B i) :=
+⟨begin
+  intros s,
+  rw hB.to_ring_filter_basis.to_add_group_filter_basis.nhds_zero_has_basis.mem_iff,
+  split,
+  { rintro ⟨-, ⟨i, rfl⟩, hi⟩,
+    exact ⟨i, trivial, hi⟩ },
+  { rintro ⟨i, -, hi⟩,
+    exact ⟨B i, ⟨i, rfl⟩, hi⟩ }
+end⟩
+
+lemma has_basis_nhds (a : A) :
+  has_basis (@nhds A hB.topology a) (λ _, true) (λ i, {b | b - a ∈ B i}) :=
+⟨begin
+  intros s,
+  rw (hB.to_ring_filter_basis.to_add_group_filter_basis.nhds_has_basis a).mem_iff,
+  simp only [exists_prop, exists_true_left],
+  split,
+  { rintro ⟨-, ⟨i, rfl⟩, hi⟩,
+    use i,
+    convert hi,
+    ext b,
+    split,
+    { intros h,
+      use [b - a, h],
+      abel },
+    { rintros ⟨c, hc, rfl⟩,
+      simpa using hc } },
+  { rintros ⟨i, hi⟩,
+    use [B i, i, rfl],
+    rw image_subset_iff,
+    rintro b b_in,
+    apply hi,
+    simpa using b_in }
+end⟩
+
+/-- Given a subgroups basis, the basis elements as open additive subgroups in the associated
+topology. -/
+def open_add_subgroup (i : ι) : @open_add_subgroup A _ hB.topology:=
+{ is_open' := begin
+    letI := hB.topology,
+    rw is_open_iff_mem_nhds,
+    intros a a_in,
+    rw (hB.has_basis_nhds a).mem_iff,
+    use [i, trivial],
+    rintros b b_in,
+    simpa using (B i).add_mem a_in b_in
+  end,
+  ..B i }
+
+@[priority 100] instance nonarchimedean : @nonarchimedean_ring A _ hB.topology :=
+begin
+  letI := hB.topology,
+  constructor,
+  intros U hU,
+  obtain ⟨i, -, hi : (B i : set A) ⊆ U⟩ := hB.has_basis_nhds_zero.mem_iff.mp hU,
+  exact ⟨hB.open_add_subgroup i, hi⟩
+end
+
+end ring_subgroups_basis
+
+variables {ι R A : Type*} [comm_ring R] [comm_ring A] [algebra R A]
+
+/-- A family of submodules in a commutative `R`-algebra `A` is a submodules basis if it satisfies
+some axioms ensuring there is a topology on `A` which is compatiable with the ring structure and
+admits this family as a basis of neighborhoods of zero. -/
+structure submodules_ring_basis (B : ι → submodule R A) : Prop :=
+(inter : ∀ i j, ∃ k, B k ≤ B i ⊓ B j)
+(left_mul : ∀ (a : A) i, ∃ j, a • B j ≤ B i)
+(mul      : ∀ i, ∃ j, (B j : set A) * B j ⊆ B i)
+
+namespace submodules_ring_basis
+
+variables {B : ι → submodule R A} (hB : submodules_ring_basis B)
+
+lemma to_ring_subgroups_basis (hB : submodules_ring_basis B) :
+  ring_subgroups_basis (λ i, (B i).to_add_subgroup) :=
+begin
+  apply ring_subgroups_basis.of_comm (λ i, (B i).to_add_subgroup) hB.inter hB.mul,
+  intros a i,
+  rcases hB.left_mul a i with ⟨j, hj⟩,
+  use j,
+  rintros b (b_in : b ∈ B j),
+  exact hj ⟨b, b_in, rfl⟩
+end
+
+/-- The topology associated to a basis of submodules in an algebra. -/
+def topology [nonempty ι] (hB : submodules_ring_basis B) : topological_space A :=
+hB.to_ring_subgroups_basis.topology
```

Comment:
Presumably we can also turn this into a `submodules_basis` (defined below), and get a topology from that. Will this lead to a diamond? (I know we aren't using TC here, but it would still be nice to avoid the diamond.)

Reply:
I added comments and an example checking this is defeq.